core/esp_spi.c

```diff
@@ -0,0 +1,212 @@
+    uint32_t count = divider >> 16;
+    if (count > 1 || divider > 1)
+    {
+        predivider = predivider > SPI_CLOCK_DIV_PRE_M + 1 ? SPI_CLOCK_DIV_PRE_M + 1 : predivider;
+        count = count > SPI_CLOCK_COUNT_NUM_M + 1 ? SPI_CLOCK_COUNT_NUM_M + 1 : count;
+        IOMUX.CONF &= ~(bus == 0 ? IOMUX_CONF_SPI0_CLOCK_EQU_SYS_CLOCK : IOMUX_CONF_SPI1_CLOCK_EQU_SYS_CLOCK);
+        SPI(bus).CLOCK = (((predivider - 1) & SPI_CLOCK_DIV_PRE_M) << SPI_CLOCK_DIV_PRE_S) |
+            (((count - 1)     & SPI_CLOCK_COUNT_NUM_M)  << SPI_CLOCK_COUNT_NUM_S) |
+            (((count / 2 - 1) & SPI_CLOCK_COUNT_HIGH_M) << SPI_CLOCK_COUNT_HIGH_S) |
+            (((count - 1)     & SPI_CLOCK_COUNT_LOW_M)  << SPI_CLOCK_COUNT_LOW_S);
+    }
+    else
+    {
+        IOMUX.CONF |= bus == 0 ? IOMUX_CONF_SPI0_CLOCK_EQU_SYS_CLOCK : IOMUX_CONF_SPI1_CLOCK_EQU_SYS_CLOCK;
+        SPI(bus).CLOCK = SPI_CLOCK_EQU_SYS_CLOCK;
+    }
+}
+
+inline static void _set_size(uint8_t bus, uint8_t bytes)
+{
+    uint16_t bits = ((uint16_t)bytes << 3) - 1;
+    const uint32_t mask = ~((SPI_USER1_MOSI_BITLEN_M << SPI_USER1_MOSI_BITLEN_S) |
+        (SPI_USER1_MISO_BITLEN_M << SPI_USER1_MISO_BITLEN_S));
+    SPI(bus).USER1 = (SPI(bus).USER1 & mask) | (bits << SPI_USER1_MOSI_BITLEN_S) |
+        (bits << SPI_USER1_MISO_BITLEN_S);
+}
+
+inline static void _wait(uint8_t bus)
+{
+    while (SPI(bus).CMD & SPI_CMD_USR)
+        ;
+}
+
+inline static void _start(uint8_t bus)
+{
+    SPI(bus).CMD |= SPI_CMD_USR;
+}
+
+inline static uint32_t _reverse_bytes(uint32_t value)
+{
+    return (value << 24) | ((value << 8) & 0x00ff0000) | ((value >> 8) & 0x0000ff00) | (value >> 24);
+}
+
+static uint32_t _spi_single_transfer (uint8_t bus, uint32_t data, uint8_t len)
+{
+    _wait(bus);
+    _set_size(bus, len);
+    spi_endianness_t e = spi_get_endianness(bus);
+    SPI(bus).W0 = e == SPI_BIG_ENDIAN ? _reverse_bytes(data) : data;
+    _start(bus);
+    _wait(bus);
+    return e == SPI_BIG_ENDIAN ? _reverse_bytes(SPI(bus).W0) : SPI(bus).W0;
+}
+
+// works properly only with little endian byte order
+static void _spi_buf_transfer (uint8_t bus, const uint8_t *out_data, uint8_t *in_data, size_t len)
+{
+    _wait(bus);
+    _set_size(bus, len);
+    memcpy((void *)&SPI(bus).W0, out_data, len);
+    _start(bus);
+    _wait(bus);
+    if (in_data)
+        memcpy(in_data, (void *)&SPI(bus).W0, len);
+}
+
+uint8_t spi_transfer_8(uint8_t bus, uint8_t data)
+{
+    return _spi_single_transfer(bus, data, sizeof(data));
+}
+
+uint16_t spi_transfer_16(uint8_t bus, uint16_t data)
+{
+    return _spi_single_transfer(bus, data, sizeof(data));
+}
+
+uint32_t spi_transfer_32(uint8_t bus, uint32_t data)
+{
+    return _spi_single_transfer(bus, data, sizeof(data));
+}
+
+void spi_transfer(uint8_t bus, const void *out_data, void *in_data, size_t len)
+{
+    if (!out_data || !len) return;
+
+    _wait(bus);
+    spi_endianness_t e = spi_get_endianness(bus);
+    spi_set_endianness(bus, SPI_LITTLE_ENDIAN);
```

Comment:
I would much prefer to see this error out (or, even better, fall back to a less-efficient approach) if it's set to `SPI_BIG_ENDIAN` rather than just silently sending data in/out in the wrong order (especially since this behavior isn't even documented!)